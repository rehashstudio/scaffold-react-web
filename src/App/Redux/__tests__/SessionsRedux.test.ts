import configureStore from 'redux-mock-store'

import SessionActions, {
  INITIAL_STATE,
  IToken,
  reducer,
  SessionTypes,
} from '../SessionRedux'

const mockStore = configureStore()
const store = mockStore()

const dummyTokens: IToken[] = [
  {
    value: 'string',
    expiresIn: 1234,
  },
]

describe('Session actions', () => {
  beforeEach(() => {
    // Runs before each test in the suite
    store.clearActions()
  })

  test('login', () => {
    store.dispatch(SessionActions.login('identifier', 'password'))
    expect(store.getActions()).toMatchSnapshot()
  })

  test('loginSuccess', () => {
    store.dispatch(SessionActions.loginSuccess(dummyTokens))
    expect(store.getActions()).toMatchSnapshot()
  })

  test('loginFailure', () => {
    store.dispatch(SessionActions.loginFailure('loginFailure error message'))
    expect(store.getActions()).toMatchSnapshot()
  })

  test('logout', () => {
    store.dispatch(SessionActions.logout(dummyTokens))
    expect(store.getActions()).toMatchSnapshot()
  })

  test('logoutSuccess', () => {
    store.dispatch(SessionActions.logoutSuccess())
    expect(store.getActions()).toMatchSnapshot()
  })

  test('logoutFailure', () => {
    store.dispatch(SessionActions.logoutFailure('logoutFailure error message'))
    expect(store.getActions()).toMatchSnapshot()
  })

  test('increment', () => {
    store.dispatch(SessionActions.increment())
    expect(store.getActions()).toMatchSnapshot()
  })

  test('decrement', () => {
    store.dispatch(SessionActions.decrement())
    expect(store.getActions()).toMatchSnapshot()
  })
})

describe('Session Reducer', () => {
  test('initial state', () => {
    const action = { type: 'DUMMY_ACTION' }
    expect(reducer(undefined, action)).toEqual(INITIAL_STATE)
  })

  test('returns correct state on login request', () => {
    const action = {
      type: SessionTypes.LOGIN,
    }
    const expectedState = INITIAL_STATE.setIn(['fetchStatus'], {
      fetching: true,
      error: undefined,
      method: 'login',
    })
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on login success', () => {
    const action = {
      type: SessionTypes.LOGIN_SUCCESS,
      tokens: dummyTokens,
    }
    const expectedState = INITIAL_STATE.setIn(['tokens'], dummyTokens)
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on login failure', () => {
    const action = {
      type: SessionTypes.LOGIN_FAILURE,
      error: 'Login Failure error',
    }
    const expectedState = INITIAL_STATE.setIn(['fetchStatus'], {
      fetching: false,
      error: action.error,
      method: 'login',
    })
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on logout request', () => {
    const action = {
      type: SessionTypes.LOGOUT,
    }
    const expectedState = INITIAL_STATE.setIn(['fetchStatus'], {
      fetching: true,
      error: undefined,
      method: 'logout',
    })
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on logout success', () => {
    const action = {
      type: SessionTypes.LOGOUT_SUCCESS,
      tokens: dummyTokens,
    }
    const expectedState = INITIAL_STATE.setIn(['tokens'], undefined)
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on logout failure', () => {
    const action = {
      type: SessionTypes.LOGOUT_FAILURE,
      error: 'Logout Failure error',
    }
    const expectedState = INITIAL_STATE.setIn(['fetchStatus'], {
      fetching: false,
      error: action.error,
      method: 'logout',
    })
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on increment', () => {
    const action = {
      type: SessionTypes.INCREMENT,
    }
    const expectedState = INITIAL_STATE.setIn(['count'], 1)
    expect(reducer(undefined, action)).toEqual(expectedState)
  })

  test('returns correct state on decrement', () => {
    const action = {
      type: SessionTypes.DECREMENT,
    }
    const expectedState = INITIAL_STATE.setIn(['count'], -1)
    expect(reducer(undefined, action)).toEqual(expectedState)
  })
})
