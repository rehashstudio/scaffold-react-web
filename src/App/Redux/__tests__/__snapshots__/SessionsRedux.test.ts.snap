// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session actions decrement 1`] = `
Array [
  Object {
    "type": "DECREMENT",
  },
]
`;

exports[`Session actions increment 1`] = `
Array [
  Object {
    "type": "INCREMENT",
  },
]
`;

exports[`Session actions login 1`] = `
Array [
  Object {
    "identifier": "identifier",
    "password": "password",
    "type": "LOGIN",
  },
]
`;

exports[`Session actions loginFailure 1`] = `
Array [
  Object {
    "error": "loginFailure error message",
    "type": "LOGIN_FAILURE",
  },
]
`;

exports[`Session actions loginSuccess 1`] = `
Array [
  Object {
    "tokens": Array [
      Object {
        "expiresIn": 1234,
        "value": "string",
      },
    ],
    "type": "LOGIN_SUCCESS",
  },
]
`;

exports[`Session actions logout 1`] = `
Array [
  Object {
    "tokens": Array [
      Object {
        "expiresIn": 1234,
        "value": "string",
      },
    ],
    "type": "LOGOUT",
  },
]
`;

exports[`Session actions logoutFailure 1`] = `
Array [
  Object {
    "error": "logoutFailure error message",
    "type": "LOGOUT_FAILURE",
  },
]
`;

exports[`Session actions logoutSuccess 1`] = `
Array [
  Object {
    "type": "LOGOUT_SUCCESS",
  },
]
`;
